tactics/declare.ml

```diff
@@ -269,34 +280,48 @@ let define_constant ~side_effect ~name cd =
       let map (body, eff) = body, eff.Evd.seff_private in
       let body = Future.chain de.proof_entry_body map in
       let de = { de with proof_entry_body = body } in
-      let de = cast_opaque_proof_entry de in
+      let de = cast_opaque_proof_entry EffectEntry de in
       [], OpaqueEntry de, false
   | ParameterEntry e ->
     [], ConstantEntry (Entries.ParameterEntry e), not (Lib.is_modtype_strict())
   | PrimitiveEntry e ->
     [], ConstantEntry (Entries.PrimitiveEntry e), false
   in
-  let kn, eff = Global.add_constant ~side_effect ~in_section name decl in
+  let kn = Global.add_constant ~in_section name decl in
   if unsafe || is_unsafe_typing_flags() then feedback_axiom();
-  kn, eff, export
+  kn, export
 
 let declare_constant ?(local = ImportDefaultBehavior) ~name ~kind cd =
   let () = check_exists name in
-  let kn, (), export = define_constant ~side_effect:PureEntry ~name cd in
+  let kn, export = define_constant ~name cd in
   (* Register the libobjects attached to the constants and its subproofs *)
   let () = List.iter register_side_effect export in
   let () = register_constant kn kind local in
   kn
 
-let declare_private_constant ?role ?(local = ImportDefaultBehavior) ~name ~kind cd =
-  let kn, eff, export = define_constant ~side_effect:EffectEntry ~name cd in
-  let () = assert (CList.is_empty export) in
+let declare_private_constant ?role ?(local = ImportDefaultBehavior) ~name ~kind de =
+  let kn, eff =
+    let de =
+      if not de.proof_entry_opaque then
+        let body, () = Future.force de.proof_entry_body in
+        let de = { de with proof_entry_body = Future.from_val (body, ()) } in
```

Comment:
Why do this force?

Reply:
Side-effects were already forced upstream, this is just making it explicit in the kernel entry.

Reply:
I don't get it. We force in cast_proof_entry so what's the point of these 2 lines?

Reply:
I think that this is due to some typing constraint in the type of the proof_entry_body field. I'll try to see if I can simplify the code nonetheless.

Reply:
Yes, indeed, I could get rid of that at the cost of changing the `proof_entry` type, but that would require messing with unrelated files. Or maybe I could have some proxy type. Does this really look important to you?

Reply:
Actually if we get the 2 PRs that make `proof_entry` totally abstract you wouldn't have more freedom to do these kind of changes.